zrml/prediction-markets/src/lib.rs

```diff
@@ -1349,6 +1350,338 @@ mod pallet {
+        /// It is slashed, if the early close request is disputed
+        /// and the `CloseMarketsEarlyOrigin`` decides to reject the early close.
+        /// The `CloseMarketsEarlyOrigin` (or root) can schedule it `now + CloseProtection...Period`
+        /// in the future. This is to prevent fat finger mistakes.
+        ///
+        /// # Weight
+        ///
+        /// Complexity: `O(n)`, where `n` is the maximum number of market ids
+        /// in `MarketIdsPerClose...` either at the old period end or new period end.
+        #[pallet::call_index(18)]
+        #[pallet::weight(
+            T::WeightInfo::schedule_early_close_as_authority(CacheSize::get(), CacheSize::get())
+                .max(T::WeightInfo::schedule_early_close_after_dispute(
+                    CacheSize::get(),
+                    CacheSize::get(),
+                ))
+                .max(T::WeightInfo::schedule_early_close_as_market_creator(
+                    CacheSize::get(),
+                    CacheSize::get(),
+                ))
+        )]
+        #[transactional]
+        pub fn schedule_early_close(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let is_authorized = T::CloseMarketEarlyOrigin::try_origin(origin.clone()).is_ok();
+            let (market, market_creator) = if !is_authorized {
+                // check if market creator below
+                let who = ensure_signed(origin)?;
+                let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+                ensure!(market.creator == who, Error::<T>::RequesterNotCreator);
+                (market, Some(who))
+            } else {
+                let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+                (market, None)
+            };
+
+            Self::ensure_market_is_active(&market)?;
+            let now_block = <frame_system::Pallet<T>>::block_number();
+            let now_time = <zrml_market_commons::Pallet<T>>::now();
+
+            let get_new_period = |block_period,
+                                  time_frame_period|
+             -> Result<
+                MarketPeriod<T::BlockNumber, MomentOf<T>>,
+                DispatchError,
+            > {
+                match &market.period {
+                    MarketPeriod::Block(range) => {
+                        let close_time = now_block.saturating_add(block_period);
+                        ensure!(close_time < range.end, Error::<T>::EarlyCloseRequestTooLate);
+                        Ok(MarketPeriod::Block(range.start..close_time))
+                    }
+                    MarketPeriod::Timestamp(range) => {
+                        let close_time = now_time.saturating_add(time_frame_period);
+                        ensure!(close_time < range.end, Error::<T>::EarlyCloseRequestTooLate);
+                        Ok(MarketPeriod::Timestamp(range.start..close_time))
```

Comment:
Shouldn't it be inclusive?

```suggestion
                        Ok(MarketPeriod::Timestamp(range.start..=close_time))
```

Reply:
Good thought! I actually copied from [here](https://github.com/zeitgeistpm/zeitgeist/blob/4e2569526fa1a8d4b9c04f7aaecc380ef8b7f09d/zrml/prediction-markets/src/lib.rs#L2441-L2445), but as you focus me on it right now, I think about it.... 

Ah.. market period doesn't allow inclusive ranges:

```
mismatched types
expected struct `std::ops::Range<<T as frame_system::Config>::BlockNumber>`
   found struct `RangeInclusive<<T as frame_system::Config>::BlockNumber>`
```

I think the effect of having inclusive ranges here is marginal. (1 block)

Reply:
How about this:

```suggestion
                        Ok(MarketPeriod::Block(range.start..close_time + 1))
```

Reply:
Ok this works, but it doesn't really has an effect. We could also extend the `CloseEarlyProtectionBlockPeriod` by one block. If I use your approach, I need to adapt the tests and benchmarks, which doesn't make any sense to me, if we could also extend the periods by one.